### PR TITLE
chore(ssa): indent NumericType into ObjectType

### PR DESCRIPTION
**Pull request by kevaundray**
# Related issue(s)

# Description

Before this PR, the NumericType and ObjectType looked like:

```rust
// node.rs
#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum ObjectType {
    NativeField,
    Boolean,
    Unsigned(u32), //bit size
    Signed(u32),   //bit size
    Pointer(ArrayId),
    Function,
    NotAnObject, //not an object
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum NumericType {
    Signed(u32),
    Unsigned(u32),
    NativeField,
}
```

Noting that Signed, Unsigned and NativeField were duplicated. This PR deduplicates the structure by doing:

```rust
pub enum ObjectType {
    Numeric(NumericType)
    Boolean,
    Pointer(ArrayId),
    Function,
    NotAnObject, //not an object
}
```

## Test additions / changes

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.
- [ ] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/code_gen.rs [hunk 1]
```diff
@@ -383,7 +383,8 @@ impl IRGenerator {
                 Value::Single(v_id)
             }
             Type::String(len) => {
-                let obj_type = ObjectType::Unsigned(8);
+                // TODO: document why this is 8
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa/code_gen.rs:386]: Minor nit here, note that if UTF8 is used then a character can be `1-4 bytes`, so we should be clear about the encoding and the layout

Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/code_gen.rs:386]: A character can be 1-N bytes really as rust's `char` type is only for Unicode scalar values. I think it is sufficient to say we encode strings as a packed array of bytes encoded in utf-8.

File: crates/noirc_evaluator/src/ssa/node.rs [hunk 2]
```diff
@@ -173,28 +173,25 @@ impl Variable {
 
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
 pub enum ObjectType {
-    NativeField,
     Boolean,
-    Unsigned(u32), //bit size
-    Signed(u32),   //bit size
+    // TODO: Why don't we put Boolean under Numeric type (u1)?
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/node.rs:177]: Might as well include the bool change as part of this PR

File: crates/noirc_evaluator/src/ssa/node.rs [hunk 3]
```diff
@@ -204,33 +201,66 @@ impl ObjectType {
...
+    pub fn unsigned_integer(bit_size: u32) -> ObjectType {
+        ObjectType::Numeric(NumericType::Unsigned(bit_size))
+    }
+    /// Returns a type that represents an signed integer
+    pub fn signed_integer(bit_size: u32) -> ObjectType {
+        ObjectType::Numeric(NumericType::Signed(bit_size))
+    }
+
+    /// Returns true, if the `ObjectType`
+    /// represents a field element
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/node.rs:230]: These comments aren't really adding much to the code. Doc comments can be useful but if they add nothing the method name and type signature don't already tell us we shouldn't pollute our code with them.

File: crates/noirc_evaluator/src/ssa/code_gen.rs (hunk 1 above)
Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa/code_gen.rs:386]: Sounds good

File: crates/noirc_evaluator/src/ssa/node.rs (hunk 2 above)
Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa/node.rs]: Done

File: crates/noirc_evaluator/src/ssa/node.rs (hunk 3 above)
Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa/node.rs:233]: I find Tiny docs to be helpful, especially when reading foreign codebases. For example, this is also something that Rust does, see [here](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#2060) 

In the link, its self explanatory what `is_empty` is doing when called on a vector, but the doc comments still spell it out, they come with examples, though I think thats something we can iterate on in the future where we see that as being appropriate.

**Comment by jfecher** [comment]
This PR looks good but needs merge conflict resolution

**Review by jfecher — APPROVED** [review]
👍

**Review by joss-aztec — APPROVED** [review]

**Pull request merged**